src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs

```diff
@@ -998,30 +969,38 @@ private ErrorRecord CreateModuleNotFoundError(string modulePath)
             return er;
         }
 
-        private Dictionary<String, List<PSModuleInfo>> GetAvailableLocallyModulesCore(string[] names, bool all, bool refresh)
+        private IEnumerable<PSModuleInfo> GetModuleForNames(string[] names, bool all, bool refresh)
         {
-            var modules = new Dictionary<string, List<PSModuleInfo>>(StringComparer.OrdinalIgnoreCase);
-            var modulePaths = ModuleIntrinsics.GetModulePath(false, Context);
-
+            HashSet<string> modulePathSet = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
             bool cleanupModuleAnalysisAppDomain = Context.TakeResponsibilityForModuleAnalysisAppDomain();
 
             try
             {
                 foreach (string path in ModuleIntrinsics.GetModulePath(false, Context))
                 {
+                    string uniquePath = path.TrimEnd(Utils.Separators.Directory);
+                    IEnumerable<PSModuleInfo> availableModules = null;
+
+                    // Ignore repeated module path.
+                    if (!modulePathSet.Add(uniquePath)) { continue; }
+
                     try
                     {
-                        var availableModules = all
-                            ? GetAllAvailableModules(path, refresh)
-                            : GetDefaultAvailableModules(names, path, refresh);
-
-                        // Add the path in $env:PSModulePath as the keys of the dictionary
-                        // If the paths are repeated, ignore the repetitions
-                        string uniquePath = path.TrimEnd(Utils.Separators.Directory);
-                        if (!modules.ContainsKey(uniquePath))
-                        {
-                            modules.Add(uniquePath, availableModules.OrderBy(m => m.Name).ToList());
-                        }
+                        var availableModuleFiles = all
+                            ? ModuleUtils.GetAllAvailableModuleFiles(uniquePath)
+                            : ModuleUtils.GetDefaultAvailableModuleFiles(uniquePath);
+
+                        availableModules = GetModulesBasedOnGivenModuleFiles(names, availableModuleFiles, refresh, all).OrderBy(
```

Comment:
This logic might actually be very useful as a method on `PSModuleInfo`. The formatter for `Get-Module -ListAvailable` [also does something like this](https://github.com/PowerShell/PowerShell/blob/bbb4f2ea84f6c4a8e8775978178826b61db20bda/src/System.Management.Automation/FormatAndOutput/DefaultFormatters/PowerShellCore_format_ps1xml.cs#L1187-L1203). I know that in PowerShellEditorServices we wanted a method or property like this.

I admit I'm also a bit averse to such a large inline delegate for a LINQ method.

Reply:
The method `GetModulesBasedOnGivenModuleFiles` was removed to further simply `GetModuleForNames`.